### ISSUE TITLE
Wal corruption after manual recovery from running out of disk

### ISSUE THREAD

**Issue by Berbe** [prometheus/prometheus#7835]
**What did you do?**
At ~13:36Z on 2020-08-22, a Prometheus instance ran out of disk.
On recovery, at ~16;15Z, disk space was restored, Prometheus started logging data points again and said data started to become usable again.

Suddenly, later on, data from ~16:15Z to 18:00Z vanished, seemingly because Prometheus detected corruption.

**What did you expect to see?**
Either:
- Safe atomic writes, leading to no corruption
- Data portion written in corrupted file being recoverable
- Data not being written in files having suffered from a write incident (like a full disk), and a new segment (ensuring it is not corrupted) started

**What did you see instead? Under which circumstances?**
Early on, as soon as disk became writeable again, data was instantly fed to the TSDB & used to be displayed in graphs again.

Some time later, this previously available data became unavailable, which is the problem. Why would that happen? It could be read before; it should not disappear like that.
~2 hours of data was erased.

**Environment**
* System information:
  Linux 4.19.0-10-amd64 x86_64

* Prometheus version:
  prometheus, version 2.20.1 (branch: HEAD, revision: 983ebb4a513302315a8117932ab832815f85e3d2)
  build user:       root@7cbd4d1c15e0
  build date:       20200805-17:26:58
  go version:       go1.14.6

* Prometheus configuration file:
```
global:
  scrape_interval: 15s
  evaluation_interval: 15s

scrape_configs:
  - job_name: 'prometheus'
    static_configs:
    - targets: ['localhost:9090']

  - job_name: system
    static_configs:
      - targets:
          - <node>:9100
```

* Logs:
```
level=info ts=2020-08-22T13:00:04.322Z caller=compact.go:495 component=tsdb msg="write block" mint=1598090400000 maxt=1598097600000 ulid=01EGB42B8BWB1JV4Z4C0GW6M8R duration=87.422504ms
level=info ts=2020-08-22T13:00:04.333Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=2.604726ms
level=error ts=2020-08-22T13:36:49.252Z caller=scrape.go:1000 component="scrape manager" scrape_pool=system target=http://172.22.0.1:9100/metrics msg="Scrape commit failed" err="write to WAL: log samples: write data/wal/00002019: no space left on device"
[...]
level=error ts=2020-08-22T16:14:49.238Z caller=scrape.go:1000 component="scrape manager" scrape_pool=system target=http://172.22.0.1:9100/metrics msg="Scrape commit failed" err="write to WAL: log samples: write data/wal/00002019: no space left on device"
level=error ts=2020-08-22T16:14:50.706Z caller=scrape.go:1000 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Scrape commit failed" err="write to WAL: log samples: write data/wal/00002019: no space left on device"
level=info ts=2020-08-22T16:15:04.349Z caller=compact.go:495 component=tsdb msg="write block" mint=1598097600000 maxt=1598104800000 ulid=01EGBF7D16DT37CJZD1KABF62X duration=118.887892ms
level=info ts=2020-08-22T16:15:04.359Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=2.419222ms
level=info ts=2020-08-22T16:15:04.378Z caller=checkpoint.go:96 component=tsdb msg="Creating checkpoint" from_segment=2016 to_segment=2017 mint=1598104800000
level=info ts=2020-08-22T16:15:04.951Z caller=head.go:884 component=tsdb msg="WAL checkpoint complete" first=2016 last=2017 duration=573.030046ms
level=info ts=2020-08-22T16:15:05.455Z caller=compact.go:441 component=tsdb msg="compact blocks" count=3 mint=1598076000000 maxt=1598097600000 ulid=01EGBF7DQY6R6YRVW6E6AX17SD sources="[01EGAPAWRDF5GZH8W6PAGK425Q 01EGAX6M0CF094KGH64YAEM370 01EGB42B8BWB1JV4Z4C0GW6M8R]" duration=497.491483ms
level=info ts=2020-08-22T16:15:06.252Z caller=compact.go:441 component=tsdb msg="compact blocks" count=3 mint=1598032800000 maxt=1598097600000 ulid=01EGBF7E84366RFRB4JRNTNA2Z sources="[01EGA1QQFG27A45CNER8T59QCP 01EGAPAWWVS73C4DSTBK906Y0R 01EGBF7DQY6R6YRVW6E6AX17SD]" duration=775.635749ms
level=info ts=2020-08-22T17:00:04.278Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=3.006126ms
level=info ts=2020-08-22T19:00:04.637Z caller=compact.go:495 component=tsdb msg="write block" mint=1598112000000 maxt=1598119200000 ulid=01EGBRNH1CXW6CHN72G83CAASD duration=339.762934ms
level=info ts=2020-08-22T19:00:04.775Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=23.660712ms
level=info ts=2020-08-22T19:00:04.836Z caller=checkpoint.go:96 component=tsdb msg="Creating checkpoint" from_segment=2018 to_segment=2019 mint=1598119200000
level=error ts=2020-08-22T19:00:05.122Z caller=db.go:685 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: read segments: corruption in segment data/wal/00002019 at 1523666: unexpected checksum c50f1836, expected 19148eb1"
level=info ts=2020-08-22T21:00:04.837Z caller=compact.go:495 component=tsdb msg="write block" mint=1598119200000 maxt=1598126400000 ulid=01EGBZH891QCQZ0R6BK6AWMMB6 duration=534.527902ms
level=info ts=2020-08-22T21:00:04.950Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=25.224651ms
level=info ts=2020-08-22T21:00:05.040Z caller=checkpoint.go:96 component=tsdb msg="Creating checkpoint" from_segment=2018 to_segment=2020 mint=1598126400000
level=error ts=2020-08-22T21:00:05.277Z caller=db.go:685 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: read segments: corruption in segment data/wal/00002019 at 1523666: unexpected checksum c50f1836, expected 19148eb1"
```

**Comment by brian-brazil** [comment]
It sounds like we might not be handling things correctly when a write to a WAL segment fails, we should ensure that if there is a partial write that fails then subsequent writes either also fail or that they work properly.

**Comment by Berbe** [comment]
In case this comment adds value, it seems data older than 4h (2x 2h?) is erased.

It seems writing to the same corrupted segment indeed continues. The proper behaviour would be to start a new one.

I would also stress that, despite the srgment being corrupted, data was exploitable from it. If compaction is not possible directly, would it be possible to rewrite readable data to an uncorrupted, temporary, segment before compacting it?
```
level=info ts=2020-08-24T15:00:04.442Z caller=compact.go:495 component=tsdb msg="write block" mint=1598270400000 maxt=1598277600000 ulid=01EGGFQGGHXTW556RA2TVB2J2R duration=201.066009ms
level=info ts=2020-08-24T15:00:04.452Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=2.791713ms
level=info ts=2020-08-24T15:00:04.488Z caller=checkpoint.go:96 component=tsdb msg="Creating checkpoint" from_segment=2018 to_segment=2034 mint=1598277600000
level=error ts=2020-08-24T15:00:04.545Z caller=db.go:685 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: read segments: corruption in segment data/wal/00002019 at 1523666: unexpected checksum c50f1836, expected 19148eb1"
level=info ts=2020-08-24T17:00:04.344Z caller=compact.go:495 component=tsdb msg="write block" mint=1598277600000 maxt=1598284800000 ulid=01EGGPK7RE3SNNWREFHM28QYX0 duration=105.925011ms
level=info ts=2020-08-24T17:00:04.355Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=2.619469ms
level=info ts=2020-08-24T17:00:04.398Z caller=checkpoint.go:96 component=tsdb msg="Creating checkpoint" from_segment=2018 to_segment=2034 mint=1598284800000
level=error ts=2020-08-24T17:00:04.565Z caller=db.go:685 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: read segments: corruption in segment data/wal/00002019 at 1523666: unexpected checksum c50f1836, expected 19148eb1"
```

**Comment by roidelapluie** [comment]
Which filesytem are you using?

**Comment by Berbe** [comment]
`ext4` inside a Docker container

**Comment by f0x52** [comment]
I just had the same issue, disk was full and now my prometheus is corrupt..
Was fixed & restarted more than a week ago, but as you can see on the gap it fails to store longer than 4 hours
https://stats.pixie.town/d/000000012/synapse?orgId=1&refresh=5m&from=now-30d&to=now

Log contains lots of 
```
level=error ts=2020-12-16T19:00:04.782Z caller=db.go:685 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: read segments: corruption in segment /var/lib/prometheus/metrics2/wal/00001406 at 41051064: unexpected checksum 203bbf7e, expected a63e7e77"
level=error ts=2020-12-16T21:00:04.686Z caller=db.go:685 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: read segments: corruption in segment /var/lib/prometheus/metrics2/wal/00001406 at 41051064: unexpected checksum 203bbf7e, expected a63e7e77
```

**Comment by codesome** [comment]
A major portion of this corruption (when the WAL record does not span multiple pages) should be fixed in https://github.com/prometheus/prometheus/pull/8125 and is available in v2.23.0. 

@f0x52 Which version of Prometheus are you running?

**Comment by f0x52** [comment]
`prometheus, version 2.20.0+ds (branch: debian/sid, revision: 2.20.0+ds-1)`
so bit outdated then (in the meantime I moved the data directory to just get it working again)

**Comment by Berbe** [comment]
That's the downside of relying on a distro known to emphasis stability over freshness.
Note that your package is not up-to-date; cf. [`prometheus` package details](https://packages.debian.org/sid/prometheus).

What you could do though, while still relying on Debian for general stability, is run upstream Prometheus' binary in a container.

**Comment by bwplotka** [comment]
Hello from the Bug Scrub!

Sounds like this source of corruption was fixed in v2.23.0. If you encounter any corruptions on the newer Prometheus versions, let us know! Thanks!
